File: backend/app/services/rating_service.py
"""
Rating Service - Business logic for ratings and reviews.
"""

from typing import List, Optional
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
import logging

from app.db.models.rating import Rating, Review
from app.schemas.rating import RatingCreate, RatingUpdate, ReviewCreate, ReviewUpdate
from app.core.exceptions import ResourceNotFoundError, ValidationError
from app.ml.sentiment.analyzer import get_sentiment_analyzer

logger = logging.getLogger(__name__)


class RatingService:
    """Service for rating and review operations."""
    
    def __init__(self, db: AsyncSession):
        self.db = db
        self.sentiment_analyzer = get_sentiment_analyzer()
    
    async def create_rating(self, user_id: str, data: RatingCreate) -> Rating:
        """Create or update movie rating."""
        # Check if rating already exists
        existing = await self.db.execute(
            select(Rating).where(
                Rating.user_id == user_id,
                Rating.movie_id == data.movie_id
            )
        )
        rating = existing.scalar_one_or_none()
        
        if rating:
            # Update existing
            for field, value in data.dict(exclude={'movie_id'}).items():
                if value is not None:
                    setattr(rating, field, value)
        else:
            # Create new
            rating = Rating(user_id=user_id, **data.dict())
            self.db.add(rating)
        
        await self.db.commit()
        await self.db.refresh(rating)
        
        # Update movie aggregate ratings
        await self._update_movie_aggregate(data.movie_id)
        
        logger.info(f"Rating created/updated: user={user_id}, movie={data.movie_id}")
        return rating
    
    async def create_review(self, user_id: str, data: ReviewCreate) -> Review:
        """Create movie review with sentiment analysis."""
        review = Review(
            user_id=user_id,
            movie_id=data.movie_id,
            rating_id=data.rating_id,
            title=data.title,
            content=data.content,
            is_spoiler=data.is_spoiler
        )
        
        self.db.add(review)
        await self.db.commit()
        await self.db.refresh(review)
        
        # Queue sentiment analysis (async)
        try:
            analysis = await self.sentiment_analyzer.analyze_review(data.content)
            review.sentiment_score = analysis["sentiment_score"]
            review.sentiment_label = analysis["sentiment_label"]
            review.sentiment_confidence = analysis["sentiment_confidence"]
            review.emotions = analysis["emotions"]
            review.aspect_sentiments = analysis["aspect_sentiments"]
            await self.db.commit()
        except Exception as e:
            logger.error(f"Sentiment analysis failed: {e}")
        
        logger.info(f"Review created: user={user_id}, movie={data.movie_id}")
        return review
    
    async def get_user_ratings(self, user_id: str, limit: int = 50) -> List[Rating]:
        """Get user's ratings."""
        result = await self.db.execute(
            select(Rating)
            .where(Rating.user_id == user_id)
            .order_by(Rating.created_at.desc())
            .limit(limit)
        )
        return list(result.scalars().all())
    
    async def get_movie_reviews(self, movie_id: int, limit: int = 20) -> List[Review]:
        """Get movie reviews."""
        result = await self.db.execute(
            select(Review)
            .where(Review.movie_id == movie_id)
            .where(Review.is_hidden == False)
            .order_by(Review.likes_count.desc())
            .limit(limit)
        )
        return list(result.scalars().all())
    
    async def _update_movie_aggregate(self, movie_id: int):
        """Update movie aggregate ratings."""
        result = await self.db.execute(
            select(
                func.avg(Rating.overall_rating).label('avg_rating'),
                func.count(Rating.id).label('count')
            ).where(Rating.movie_id == movie_id)
        )
        row = result.first()
        # Would update Movie model here
        logger.info(f"Movie aggregate updated: movie={movie_id}, avg={row[0]}, count={row[1]}")

File: frontend/src/services/rating.service.ts
/**
 * Rating and Review Service
 * Handles all rating and review operations
 */

import { apiClient } from './api.client';

// Types
export interface CreateRatingRequest {
  movie_id: number;
  overall_rating: number;
  plot_rating?: number;
  acting_rating?: number;
  cinematography_rating?: number;
  soundtrack_rating?: number;
}

export interface Rating {
  id: number;
  movie_id: number;
  overall_rating: number;
  plot_rating?: number;
  acting_rating?: number;
  cinematography_rating?: number;
  soundtrack_rating?: number;
  created_at: string;
  movie?: import('@/types').Movie;
}

export interface CreateReviewRequest {
  movie_id: number;
  rating_id: number;
  title?: string;
  content: string;
  is_spoiler: boolean;
}

export interface Review {
  id: number;
  movie_id: number;
  user_id: string;
  title?: string;
  content: string;
  sentiment_score?: number;
  sentiment_label?: string;
  is_spoiler: boolean;
  likes_count: number;
  created_at: string;
  user?: {
    username: string;
    avatar_url?: string;
  };
}

export const ratingService = {
  /**
   * Rate a movie (create or update rating)
   */
  async rateMovie(data: CreateRatingRequest): Promise<Rating> {
    return await apiClient.post<Rating>('/ratings/rate', data);
  },

  /**
   * Get current user's ratings
   */
  async getMyRatings(page: number = 1, pageSize: number = 20): Promise<Rating[]> {
    return await apiClient.get<Rating[]>('/ratings/my-ratings', {
      params: { page, page_size: pageSize },
    });
  },

  /**
   * Create a review for a movie
   */
  async createReview(data: CreateReviewRequest): Promise<Review> {
    return await apiClient.post<Review>('/ratings/review', data);
  },

  /**
   * Get reviews for a movie
   */
  async getMovieReviews(
    movieId: number,
    page: number = 1,
    pageSize: number = 20
  ): Promise<Review[]> {
    return await apiClient.get<Review[]>(`/ratings/movie/${movieId}/reviews`, {
      params: { page, page_size: pageSize },
    });
  },

  /**
   * Like a review
   */
  async likeReview(reviewId: number): Promise<void> {
    await apiClient.post(`/ratings/review/${reviewId}/like`);
  },

  /**
   * Delete a rating
   */
  async deleteRating(ratingId: number): Promise<void> {
    await apiClient.delete(`/ratings/${ratingId}`);
  },
};
